pkg/distributor/distributor.go

```diff
@@ -1460,32 +1460,40 @@ func (d *Distributor) labelValuesCardinality(ctx context.Context, labelNames []m
 		return nil, err
 	}
 
-	// Make sure we get a successful response from all the ingesters
-	replicationSet.MaxErrors = 0
-	replicationSet.MaxUnavailableZones = 0
+	// If we have a single zone, we require all ingesters to respond.
+	if replicationSet.ZoneCount() == 1 {
+		replicationSet.MaxErrors = 0
+	}
 
 	cardinalityConcurrentMap := &labelValuesCardinalityConcurrentMap{
-		cardinalityMap: map[string]map[string]uint64{},
+		cardinalityMapByZone: map[string]map[string]map[string]uint64{},
 	}
 
 	labelValuesReq, err := toLabelValuesCardinalityRequest(labelNames, matchers)
 	if err != nil {
 		return nil, err
 	}
 
-	_, err = d.forReplicationSet(ctx, replicationSet, func(ctx context.Context, client ingester_client.IngesterClient) (interface{}, error) {
+	_, err = ring.DoUntilQuorum[struct{}](ctx, replicationSet, func(ctx context.Context, desc *ring.InstanceDesc) (struct{}, error) {
+		poolClient, err := d.ingesterPool.GetClientFor(desc.Addr)
+		if err != nil {
+			return struct{}{}, err
+		}
+
+		client := poolClient.(ingester_client.IngesterClient)
+
 		stream, err := client.LabelValuesCardinality(ctx, labelValuesReq)
 		if err != nil {
-			return nil, err
+			return struct{}{}, err
 		}
 		defer func() { _ = stream.CloseSend() }()
 
-		return nil, cardinalityConcurrentMap.processLabelValuesCardinalityMessages(stream)
-	})
+		return struct{}{}, cardinalityConcurrentMap.processLabelValuesCardinalityMessages(desc.Zone, stream)
```

Comment:
Have you considered instead accumulating the streams in a slice and looping over them after `DoUntilQuorum` returns? It feels like that was the intended usage of `DoUntilQuorum` as opposed to consuming the result within the function.

Reply:
ah, that sounds great. Maybe a comment could help. This is merged now, so i don't insist you add it